src/index.c

```diff
@@ -1087,6 +1163,76 @@ static int NI_ReadSorted(void *ctx, RSIndexResult **hit) {
   return INDEXREAD_OK;
 }
 
+/* Read from a NOT iterator - Optimized version, utilizing the `existing docs`
+ * inverted index. This is applicable only if the only or leftmost node of a
+ * query is a NOT node. We simply read until max docId, skipping docIds that
+ * exist in the child */
+static int NI_ReadSorted_O(void *ctx, RSIndexResult **hit) {
+  NotContext *nc = ctx;
+  if (nc->lastDocId > nc->maxDocId) {
+    IITER_SET_EOF(&nc->base);
+    return INDEXREAD_EOF;
+  }
+
+  RSIndexResult *cr = NULL;
+  // if we have a child, get the latest result from the child
+  cr = IITER_CURRENT_RECORD(nc->child);
+
+  if (cr == NULL || cr->docId == 0) {
+    nc->child->Read(nc->child->ctx, &cr);
+  }
+
+  // Advance the embedded wildcard iterator
+  RSIndexResult *wcii_res = NULL;
+  int wcii_rc = nc->wcii->Read(nc->wcii->ctx, &wcii_res);
+
+  if (wcii_rc == INDEXREAD_EOF) {
+    // If the wildcard iterator hit EOF, we're done
+    IITER_SET_EOF(&nc->base);
+    return INDEXREAD_EOF;
+  }
+  nc->base.current->docId = wcii_res->docId;
+
+  // If there is no child result, or the child result is ahead of the wildcard
+  // iterator result, we wish to return the current docId.
+  if (cr == NULL || cr->docId > wcii_res->docId || !IITER_HAS_NEXT(nc->child)) {
+    goto ok;
+  }
+
+  while (cr->docId <= wcii_res->docId) {
+    wcii_rc = nc->wcii->Read(nc->wcii->ctx, &wcii_res);
```

Comment:
If `cr->docId < wcii_res->docId` (is it even possible?), do we want to read from the wildcard? Don't we need to advance it only if we reach it? (`cr->docId == wcii_res->docId`)

Reply:
Great question.
I think we can reach a situation in which `cr->docId < wcii_res->docId` if the GC was able to clean a docId from the `existingDocs` II but not from the child II since it was in the last block of the child II but not the last in the `existingDocs` II.
In such a case, the doc was deleted of course so we should just read from the child again. Then we may get a docId that is not equal to the wildcard docId, but larger, such that we should return the current docId.
I don't remember why I was convinced that this solves this case, will add an inner loop for the case that we have a smaller Id.